classes/app/setting/admin_settings.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+namespace assignsubmission_pxaiwriter\app\setting;
+
+
+use assignsubmission_pxaiwriter\app\interfaces\factory as base_factory;
+use stdClass;
+
+/* @codeCoverageIgnoreStart */
+defined('MOODLE_INTERNAL') || die();
+/* @codeCoverageIgnoreEnd */
+
+class admin_settings implements interfaces\admin_settings
+{
+    private base_factory $factory;
+    private ?object $config;
+
+    public function __construct(base_factory $factory)
+    {
+        $this->factory = $factory;
+        $this->config = $this->factory->moodle()->get_config_instance() ?? new stdClass();
+    }
+
+    public function get_default(): bool
+    {
+        return $this->config->default ?? true;
+    }
+
+    public function get_presence_penalty(): float
+    {
+        return $this->config->presence_penalty ?? 0;
+    }
+
+    public function get_frequency_penalty(): float
+    {
+        return $this->config->frequency_penalty ?? 0;
+    }
+
+    public function get_top_p(): float
+    {
+        return $this->config->top_p ?? 1;
+    }
+
+    public function get_max_tokens(): int
+    {
+        return $this->config->max_tokens ?? 256;
+    }
+
+    public function get_temperature(): float
+    {
+        return $this->config->temperature ?? 0.7;
+    }
+
+    public function get_model(): string
+    {
+        return $this->config->model ?? 'text-davinci-002';
+    }
+
+    public function get_authorization(): string
+    {
+        return $this->config->authorization ?? '';
+    }
+
+    public function get_url(): string
+    {
+        return $this->config->url ?? 'https://api.openai.com/v1/completions';
```

Comment:
I think we should throw en exception instead, if no API URL has been configured.

Reply:
Something moodle doesn't got save the configuration and it leave empty even though it has a default value.
So this is just an insurance for it to return a default value.

If it has no default value, then it will make sense to throw it.